collectors/plugins.d/README.md

```diff
@@ -362,6 +364,80 @@ The `source` is an integer field that can have the following values:
 
 `CLABEL_COMMIT` indicates that all labels were defined and the chart can be updated.
 
+#### FUNCTION
+
+> FUNCTION [GLOBAL] "name and parameters of the function" timeout "help string for users"
+
+A function can be used by users to ask for more information from the collector. Netdata maintains a registry of functions in 2 levels:
+
+- per node
+- per chart
+
+Both node and chart functions are exactly the same, but chart functions allow Netdata to relate functions with charts and therefore present a context sensitive menu of functions related to the chart the user is using.
+
+A function is identified by a string. The allowed characters in the function definition are:
+
+| Character         | Symbol | In Functions |
+|-------------------|:------:|:------------:|
+| UTF-8 character   | UTF-8  |     keep     |
+| Lower case letter | [a-z]  |     keep     |
+| Upper case letter | [A-Z]  |     keep     |
+| Digit             | [0-9]  |     keep     |
+| Underscore        |   _    |     keep     |
+| Comma             |   ,    |     keep     |
+| Minus             |   -    |     keep     |
+| Period            |   .    |     keep     |
+| Colon             |   :    |     keep     |
+| Slash             |   /    |     keep     |
+| Space             |  ' '   |     keep     |
+| Semicolon         |   ;    |      :       |
+| Equal             |   =    |      :       |
+| Backslash         |   \    |      /       |
+| Anything else     |        |      _       |
+
+Uses can get a list of all the registered functions using the `/api/v1/functions` end point of Netdata.
+
+Users can call functions using the `/api/v1/function` end point of Netdata.
+Once a function is called, the plugin will receive at its standard input a command that looks like this:
+
+> FUNCTION transaction_id timeout "name and parameters of the function"
+
+The plugin is expected to parse and validate `name and parameters of the function`. Netdata allows users to edit this string, append more parameters or even change the ones the plugin originally exposed. To minimize the security risk, Netdata guarantees that only the characters shown above are accepted in function definitions, but still the plugin should carefully inspect the `name and parameters of the function` to ensure that it is valid and not harmful.
```

Comment:
`name and parameters of the function` - how it should look? Documentation seems unclear to me. If it is supposed to be arbitrary string to be interpreted at plugins own free will shouldn't we call it something like `command`?
If the format is enforced to `"fnc_name [parameters]*"` shouldnt the function name be a separate parameter on the line? (not together in string with params)

Reply:
This should be a contract between UI and plugins. This is why I have not explained it more currently.
I believe it should be something like:

```
FUNCTION [parameter] [parameter]
```

When each `parameter` should be `key[:value[,value]]`.

So, any `key` is accepted. But also `key:value` or even `key:value1,value2,value3`.

The way this works today, is that the plugin should register a function with the minimum number of parameters required to execute the function. For example, a plugin may say:

```
PostgresqlSlowQuery db:sales
```

But then the plugin could allow the user to provide additional parameters, like `table:revenue` or even `last:100`.

Based on the discussions we will have with the UI, we may also support a number parameters to help the UI be more performant, like `columns:col1,col2,col3,col4` to request fewer of the available columns, or even `sort:column:ascending rows:100` to have the plugin return the top 100 rows after sorting the table on `column` ascending.

So, all these remain to be fine decided.